services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/geometria/GeometriatServiceImpl.kt

```diff
@@ -2,68 +2,72 @@ package fi.hel.haitaton.hanke.geometria
-                geometriat.id = oldGeometriat!!.id
-                geometriat.createdByUserId = oldGeometriat.createdByUserId
-                geometriat.createdAt = oldGeometriat.createdAt
-                geometriat.modifiedByUserId = username
-                geometriat.modifiedAt = now
-                geometriat.version = oldGeometriat.version!! + 1
-                hankeGeometriaDao.deleteGeometriat(geometriat)
-                logger.info { "Deleted geometries ${geometriat.id}" }
-                geometriat
-            }
-            oldGeometriat == null -> {
-                // CREATE
-                geometriat.createdByUserId = username
-                geometriat.createdAt = now
-                geometriat.modifiedByUserId = null
-                geometriat.modifiedAt = null
-                geometriat.version = 0
-                val created = hankeGeometriaDao.createGeometriat(geometriat)
-                logger.info { "Created new geometries ${created.id}" }
-                created
-            }
-            else -> {
-                // UPDATE
-                if (oldGeometriat.version == null) {
-                    error("There is an old Geometriat ${oldGeometriat.id} but it has no 'version'")
-                } else {
-                    oldGeometriat.version = oldGeometriat.version!! + 1
-                }
-                oldGeometriat.modifiedByUserId = username
-                oldGeometriat.modifiedAt = now
-                oldGeometriat.featureCollection = geometriat.featureCollection
-                hankeGeometriaDao.updateGeometriat(oldGeometriat)
-                logger.info { "Updated geometries ${oldGeometriat.id}" }
-                oldGeometriat
-            }
+            oldGeometriat != null && !hasFeatures -> deleteGeometriat(oldGeometriat)
+            oldGeometriat == null -> createGeometriat(geometriat)
+            else -> updateGeometriat(oldGeometriat, geometriat.featureCollection!!)
         }
     }
 
+    override fun createGeometriat(geometriat: HasFeatures): Geometriat {
+        val created =
+            Geometriat(
+                    featureCollection = geometriat.featureCollection,
+                    createdByUserId = currentUserId(),
+                    createdAt = ZonedDateTime.now(TZ_UTC),
+                    modifiedByUserId = null,
+                    modifiedAt = null,
+                    version = 0,
+                )
+                .let(hankeGeometriaDao::createGeometriat)
+
+        logger.info { "Created new geometries ${created.id}" }
+        return created
+    }
+
+    private fun deleteGeometriat(geometriat: Geometriat): Geometriat? {
+        hankeGeometriaDao.deleteGeometriat(geometriat)
+        logger.info { "Deleted geometries ${geometriat.id}" }
+        return null
+    }
+
+    private fun updateGeometriat(
+        geometriat: Geometriat,
+        newFeatures: FeatureCollection
+    ): Geometriat {
+        if (geometriat.version == null) {
```

Comment:
Suggestion. No need for else branch, and removing it makes the if foldable to an elvis operator.

```
    private fun updateGeometriat(
        geometriat: Geometriat,
        newFeatures: FeatureCollection
    ): Geometriat {
        val version =
            geometriat.version
                ?: error("There is an old Geometriat ${geometriat.id} but it has no 'version'")
        geometriat.version = version + 1
        geometriat.modifiedByUserId = currentUserId()
        geometriat.modifiedAt = ZonedDateTime.now(TZ_UTC)
        geometriat.featureCollection = newFeatures
        hankeGeometriaDao.updateGeometriat(geometriat)
        logger.info { "Updated geometries ${geometriat.id}" }
        return geometriat
    }
```

Reply:
I made the version non-nullable so that we can avoid the complication altogether.

I'm not sure why adding a new hankealue to an existing hanke still works, but it does.